packages/admin-panel-server/src/viz-builder/extractors/DashboardVisualisationExtractor.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup } from '@tupaia/utils';
+
+import { DashboardVisualisationObject } from '../types';
+import { omitBy, isNil } from 'lodash';
+
+export abstract class DashboardVisualisationExtractor {
+  private readonly visualisationObject: DashboardVisualisationObject;
+
+  private validationSchema: yup.ObjectSchema;
+
+  private dataValidationSchema: yup.ObjectSchema;
+
+  private presentationValidationSchema: yup.ObjectSchema;
+
+  constructor(visualisationObject: DashboardVisualisationObject) {
+    this.visualisationObject = visualisationObject;
+  }
+
+  protected getValidationSchema() {
+    if (!this.validationSchema) {
+      this.validationSchema = yup.object().shape({
+        code: yup.string().required(),
+        name: yup.string().required(),
+        permissionGroup: yup.string().required(),
+      });
+    }
+    
+    return this.validationSchema;
+  }
+
+  protected getDataValidationSchema() {
+    if (!this.dataValidationSchema) {
+      // Even if just draft version, data/report should at least have dataElements or dataGroups defined
+      this.dataValidationSchema = yup.object().shape({
+        dataElements: yup
+          .array()
+          .of(yup.string())
+          .when('dataGroups', { // dataGroups is required if there are no dataElements 
+            is: (dataGroups: string[]) => !dataGroups || dataGroups.length === 0,
+            then: yup.array().of(yup.string()).required().min(1),
+            otherwise: yup.array().of(yup.string()),
+          }),
+        dataGroups: yup
+          .array()
+          .of(yup.string())
+          .when('dataElements', { // dataElements is required if there are no dataGroups 
+            is: (dataElements: string[]) => !dataElements || dataElements.length === 0,
+            then: yup.array().of(yup.string()).required().min(1),
+            otherwise: yup.array().of(yup.string()),
+          }),
```

Comment:
When saving a draft version, I think we still need at to have some minimum requirements so that we just don't save whatever and avoid some column constraints in the database